heartbeat/monitor/monitor.go

```diff
@@ -0,0 +1,200 @@
+		pubKeyConverter:               args.PubKeyConverter,
+		marshaller:                    args.Marshaller,
+		peerShardMapper:               args.PeerShardMapper,
+		maxDurationPeerUnresponsive:   args.MaxDurationPeerUnresponsive,
+		hideInactiveValidatorInterval: args.HideInactiveValidatorInterval,
+		shardId:                       args.ShardId,
+		numInstances:                  make(map[string]uint64, 0),
+	}, nil
+}
+
+func checkArgs(args ArgHeartbeatV2Monitor) error {
+	if check.IfNil(args.Cache) {
+		return heartbeat.ErrNilCacher
+	}
+	if check.IfNil(args.PubKeyConverter) {
+		return heartbeat.ErrNilPubkeyConverter
+	}
+	if check.IfNil(args.Marshaller) {
+		return heartbeat.ErrNilMarshaller
+	}
+	if check.IfNil(args.PeerShardMapper) {
+		return heartbeat.ErrNilPeerShardMapper
+	}
+	if args.MaxDurationPeerUnresponsive < minDuration {
+		return fmt.Errorf("%w on MaxDurationPeerUnresponsive, provided %d, min expected %d",
+			heartbeat.ErrInvalidTimeDuration, args.MaxDurationPeerUnresponsive, minDuration)
+	}
+	if args.HideInactiveValidatorInterval < minDuration {
+		return fmt.Errorf("%w on HideInactiveValidatorInterval, provided %d, min expected %d",
+			heartbeat.ErrInvalidTimeDuration, args.HideInactiveValidatorInterval, minDuration)
+	}
+
+	return nil
+}
+
+// GetHeartbeats returns the heartbeat status
+func (monitor *heartbeatV2Monitor) GetHeartbeats() []data.PubKeyHeartbeat {
+	monitor.numInstances = make(map[string]uint64, 0)
+
+	pids := monitor.cache.Keys()
+
+	heartbeatsV2 := make([]data.PubKeyHeartbeat, 0)
+	for idx := 0; idx < len(pids); idx++ {
+		pid := pids[idx]
+		peerId := core.PeerID(pid)
+		hb, ok := monitor.cache.Get(pid)
+		if !ok {
+			log.Debug("could not get data from cache for pid", "pid", peerId.Pretty())
+			continue
+		}
+
+		heartbeatData, err := monitor.parseMessage(peerId, hb)
+		if err != nil {
+			log.Debug("could not parse message for pid", "pid", peerId.Pretty(), "error", err.Error())
+			continue
+		}
+
+		heartbeatsV2 = append(heartbeatsV2, heartbeatData)
+	}
+
+	for idx := range heartbeatsV2 {
+		pk := heartbeatsV2[idx].PublicKey
+		heartbeatsV2[idx].NumInstances = monitor.numInstances[pk]
+	}
+
+	sort.Slice(heartbeatsV2, func(i, j int) bool {
+		return strings.Compare(heartbeatsV2[i].PublicKey, heartbeatsV2[j].PublicKey) < 0
+	})
+
+	return heartbeatsV2
+}
+
+func (monitor *heartbeatV2Monitor) parseMessage(pid core.PeerID, message interface{}) (data.PubKeyHeartbeat, error) {
+	pubKeyHeartbeat := data.PubKeyHeartbeat{}
+
+	heartbeatV2, ok := message.(heartbeat.HeartbeatV2)
+	if !ok {
+		return pubKeyHeartbeat, process.ErrWrongTypeAssertion
+	}
+
+	payload := heartbeat.Payload{}
+	err := monitor.marshaller.Unmarshal(&payload, heartbeatV2.Payload)
+	if err != nil {
+		return pubKeyHeartbeat, err
+	}
+
+	peerInfo := monitor.peerShardMapper.GetPeerInfo(pid)
+
+	crtTime := time.Now()
+	messageAge := monitor.getMessageAge(crtTime, payload.Timestamp)
+	stringType := string(rune(peerInfo.PeerType))
```

Comment:
do we really need the conversion to rune?

Reply:
not needed, fixed